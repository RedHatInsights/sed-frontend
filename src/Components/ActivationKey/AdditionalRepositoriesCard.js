import React from 'react';
import { useState } from 'react';
import propTypes from 'prop-types';
import {
  Text,
  TextContent,
  TextVariants,
  Card,
  CardHeader,
  CardTitle,
  CardBody,
  Title,
} from '@patternfly/react-core';
import AdditionalRepositoriesTable from '../AdditionalRepositoriesTable';
import useAvailableRepositories from '../../hooks/useAvailableRepositories';
import AddAdditionalRepositoriesButton from '../ActivationKey/AddAdditionalRepositoriesButton';
import AddAdditionalRepositoriesModal from '../Modals/AddAdditionalRepositoriesModal';

const AdditionalRepositoriesCard = (props) => {
  const { activationKey } = props;

  const {
    data: availableRepositories,
    isLoading,
    error,
  } = useAvailableRepositories(activationKey.name);

  const [
    isEditAdditionalRepositoriesModalOpen,
    setisEditAdditionalRepositoriesModalOpen,
  ] = useState(false);

  const handleEditAdditionalRepositoriesToggle = () => {
    setisEditAdditionalRepositoriesModalOpen(
      !isEditAdditionalRepositoriesModalOpen
    );
  };

  return (
    <Card>
      <CardHeader>
        <CardTitle>
          <Title headingLevel="h2"> Additional repositories </Title>
        </CardTitle>
      </CardHeader>
      <CardBody>
        <TextContent>
          <Text component={TextVariants.p}>
            The core repositories for your operating system version, for example
            BaseOS and AppStream, are always enabled and do not need to be
            explicitly added to the activation key.
          </Text>
          <AddAdditionalRepositoriesButton
            onClick={handleEditAdditionalRepositoriesToggle}
          />
          <AddAdditionalRepositoriesModal
            isOpen={isEditAdditionalRepositoriesModalOpen}
            handleModalToggle={handleEditAdditionalRepositoriesToggle}
            keyName={activationKey.name}
            repositories={availableRepositories}
            isLoading={isLoading}
            error={error}
          />
        </TextContent>
        <AdditionalRepositoriesTable
          repositories={activationKey.additionalRepositories}
          name={activationKey.name}
        />
      </CardBody>
    </Card>
  );
};

AdditionalRepositoriesCard.propTypes = {
  activationKey: propTypes.object,
  isLoading: propTypes.bool,
  error: propTypes.bool,
};

export default AdditionalRepositoriesCard;
