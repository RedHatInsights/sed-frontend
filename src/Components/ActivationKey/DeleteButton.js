import React from 'react';
import { useQueryClient } from 'react-query';
import { Button } from '@patternfly/react-core';
import PropTypes from 'prop-types';

const DeleteButton = (props) => {
  const queryClient = useQueryClient();
  const user = queryClient.getQueryData('user');
  const isDisabled = () => {
    return !user.rbacPermissions.canWriteActivationKeys;
  };
  const { onClick } = props;
  return (
    <React.Fragment>
      <Button onClick={onClick} isDisabled={isDisabled()} variant="secondary">
        Delete
      </Button>
    </React.Fragment>
  );
};

DeleteButton.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default DeleteButton;
