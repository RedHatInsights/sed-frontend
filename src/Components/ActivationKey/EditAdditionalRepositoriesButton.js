import React from 'react';
import PropTypes from 'prop-types';
import { WriteOnlyButton } from '../WriteOnlyButton';

const AddAdditionalRepositoriesButton = ({ onClick }) => {
  return (
    <React.Fragment>
      <WriteOnlyButton
        onClick={onClick}
        variant="secondary"
        style={{ margin: 15, marginLeft: 0 }}
      >
        Add repositories
      </WriteOnlyButton>
    </React.Fragment>
  );
};

AddAdditionalRepositoriesButton.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default AddAdditionalRepositoriesButton;
