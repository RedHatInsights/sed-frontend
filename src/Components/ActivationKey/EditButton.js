import React from 'react';
import PropTypes from 'prop-types';
import { WriteOnlyButton } from '../WriteOnlyButton';

const EditButton = ({ onClick }) => {
  return (
    <React.Fragment>
      <WriteOnlyButton onClick={onClick} variant="secondary">
        Edit
      </WriteOnlyButton>
    </React.Fragment>
  );
};

EditButton.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default EditButton;
