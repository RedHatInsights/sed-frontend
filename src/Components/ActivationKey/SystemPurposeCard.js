import React from 'react';
import {
  TextContent,
  Card,
  CardHeader,
  CardTitle,
  CardBody,
  TextList,
  Text,
  TextListVariants,
  TextListItem,
  TextListItemVariants,
  Title,
  } from '@patternfly/react-core';
import EditButton from './EditButton';
import propTypes from 'prop-types';
import ActivationKeysDocsPopover from '../ActivationKeysDocsPopover';
import NoAccessPopover from '../NoAccessPopover';
import { useQueryClient } from '@tanstack/react-query';

const SystemPurposeCard = (props) => {
  const { activationKey, actionHandler } = props;
  const notDefinedText = 'Not defined';
  const queryClient = useQueryClient();
  const user = queryClient.getQueryData(['user']);

  const ButtonWrapper = () => {
    return <EditButton onClick={actionHandler} />;
  };

  const docsPopoverContent = (
    <TextContent>
      <Text>
        System purpose values are used by the subscriptions service to help
        filter and identify hosts. Setting values for these attributes is
        optional, but doing so ensures that subscriptions reporting accurately
        reflects the system.
      </Text>
    </TextContent>
  );
  return (
    <Card>
      <CardHeader actions={{ actions: <>{user.rbacPermissions.canWriteActivationKeys ? (
            <ButtonWrapper />
          ) : (
            <NoAccessPopover content={ButtonWrapper} />
          )}</>, hasNoOffset: false, className: undefined}} >
        <CardTitle>
          <Title headingLevel="h2">
            System Purpose{' '}
            <ActivationKeysDocsPopover
              popoverContent={docsPopoverContent}
              position="top"
            />{' '}
          </Title>
        </CardTitle>
        
      </CardHeader>
      <CardBody>
        <TextContent>
          <TextList component={TextListVariants.dl}>
            <TextListItem component={TextListItemVariants.dt}>
              Role
            </TextListItem>
            <TextListItem component={TextListItemVariants.dd}>
              {activationKey && activationKey.role
                ? activationKey.role
                : notDefinedText}
            </TextListItem>
            <TextListItem component={TextListItemVariants.dt}>SLA</TextListItem>
            <TextListItem component={TextListItemVariants.dd}>
              {activationKey && activationKey.serviceLevel
                ? activationKey.serviceLevel
                : notDefinedText}
            </TextListItem>
            <TextListItem component={TextListItemVariants.dt}>
              Usage
            </TextListItem>
            <TextListItem component={TextListItemVariants.dd}>
              {activationKey && activationKey.usage
                ? activationKey.usage
                : notDefinedText}
            </TextListItem>
          </TextList>
        </TextContent>
      </CardBody>
    </Card>
  );
};

SystemPurposeCard.propTypes = {
  activationKey: propTypes.object,
  actionHandler: propTypes.func,
};

export default SystemPurposeCard;
