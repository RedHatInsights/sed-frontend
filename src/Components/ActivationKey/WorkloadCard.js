import React from 'react';
import {
  TextContent,
  Text,
  Card,
  CardHeader,
  CardTitle,
  CardBody,
  TextList,
  TextListVariants,
  TextListItem,
  TextListItemVariants,
  Title,
  } from '@patternfly/react-core';
import propTypes from 'prop-types';
import ActivationKeysDocsPopover from '../ActivationKeysDocsPopover';
import EditButton from './EditButton';

const WorkloadCard = (props) => {
  const { activationKey, actionHandler } = props;
  const notDefinedText = 'Not defined';
  const docsPopoverContent = (
    <TextContent>
      <Text>
        A release version enables you to configure your system to use a specific
        minor release of Red Hat Enterprise Linux. Setting a release version is
        useful if you are using an extended release of software, such as
        Extended Update Support. Most users will not set a release version.
      </Text>
    </TextContent>
  );
  return (
    <Card style={{ minHeight: '100%' }}>
      <CardHeader actions={{ actions: <><EditButton onClick={actionHandler} /></>, hasNoOffset: false, className: undefined}} >
        <CardTitle>
          <Title headingLevel="h2">
            Workload{' '}
            <ActivationKeysDocsPopover
              popoverContent={docsPopoverContent}
              position="top"
            />
          </Title>
        </CardTitle>
        
      </CardHeader>
      <CardBody>
        <TextContent>
          <TextList component={TextListVariants.dl}>
            <TextListItem component={TextListItemVariants.dt}>
              Release version
            </TextListItem>
            <TextListItem component={TextListItemVariants.dd}>
              {activationKey && activationKey.releaseVersion
                ? activationKey.releaseVersion
                : notDefinedText}
            </TextListItem>
          </TextList>
        </TextContent>
      </CardBody>
    </Card>
  );
};

WorkloadCard.propTypes = {
  activationKey: propTypes.object,
  actionHandler: propTypes.func,
};

export default WorkloadCard;
