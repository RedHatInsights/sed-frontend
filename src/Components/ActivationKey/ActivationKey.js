import React, { useState } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import Breadcrumbs from '../shared/breadcrumbs';
import {
  Text,
  TextContent,
  TextVariants,
  Grid,
  GridItem,
  Gallery,
  GalleryItem,
  Level,
  LevelItem,
} from '@patternfly/react-core';
import { Main } from '@redhat-cloud-services/frontend-components/Main';
import {
  PageHeader,
  PageHeaderTitle,
} from '@redhat-cloud-services/frontend-components/PageHeader';
import AdditionalRepositoriesCard from './AdditionalRepositoriesCard';
import useActivationKey from '../../hooks/useActivationKey';
import Loading from '../LoadingState/Loading';
import SystemPurposeCard from './SystemPurposeCard';
import WorkloadCard from './WorkloadCard';
import DeleteButton from './DeleteButton';
import DeleteActivationKeyConfirmationModal from '../Modals/DeleteActivationKeyConfirmationModal';
import EditActivationKeyModal from '../Modals/EditActivationKeyModal';
import NoAccessPopover from '../NoAccessPopover';
import { useQueryClient } from '@tanstack/react-query';
import { EditReleaseVersionModal } from '../Modals/EditReleaseVersionModal';
import useReleaseVersions from '../../hooks/useReleaseVersions';

const ActivationKey = () => {
  const navigate = useNavigate();
  const queryClient = useQueryClient();
  const user = queryClient.getQueryData(['user']);
  const { id } = useParams();
  const breadcrumbs = [
    { title: 'Activation Keys', to: '../activation-keys' },
    { title: id, isActive: true },
  ];
  const {
    isLoading: isKeyLoading,
    error: keyError,
    data: activationKey,
  } = useActivationKey(id);
  const { isLoading: areReleaseVersionsLoading, data: releaseVersions } =
    useReleaseVersions();

  const description =
    'View and edit details and repositories for this activation key.';
  const [isDeleteActivationKeyModalOpen, setIsDeleteActivationKeyModalOpen] =
    useState(false);
  const [isEditActivationKeyModalOpen, setIsEditActivationKeyModalOpen] =
    useState(false);
  const [isEditReleaseVersionModalOpen, setIsEditReleaseVersionModalOpen] =
    useState(false);
  const handleDeleteActivationKeyModalToggle = (keyDeleted) => {
    setIsDeleteActivationKeyModalOpen(!isDeleteActivationKeyModalOpen);
    if (keyDeleted === true) {
      navigate('../activation-keys');
    }
  };

  const handleEditActivationKeyModalToggle = () => {
    setIsEditActivationKeyModalOpen(!isEditActivationKeyModalOpen);
  };

  const handleEditReleaseVersionModalToggle = () => {
    setIsEditReleaseVersionModalOpen(!isEditReleaseVersionModalOpen);
  };

  const editModalDescription =
    'System purpose values are used by the subscriptions service to help filter and identify hosts. Setting values for these attributes is optional, but doing so ensures that subscriptions reporting accurately reflects the system. Only those values available to your account are shown.';

  return (
    <React.Fragment>
      <PageHeader>
        <Level>
          <LevelItem>
            <Breadcrumbs {...breadcrumbs} />
            <PageHeaderTitle title={id} />
            <TextContent>
              <Text component={TextVariants.p}>{description}</Text>
            </TextContent>
          </LevelItem>
          <LevelItem>
            {user.rbacPermissions.canWriteActivationKeys ? (
              <DeleteButton onClick={handleDeleteActivationKeyModalToggle} />
            ) : (
              <NoAccessPopover content={DeleteButton} />
            )}
          </LevelItem>
        </Level>
      </PageHeader>
      {isKeyLoading && !keyError ? (
        <Loading />
      ) : (
        <React.Fragment>
          <Main>
            <Grid hasGutter>
              <GridItem span={12}>
                <Gallery
                  hasGutter
                  minWidths={{
                    default: '40%',
                  }}
                >
                  <GalleryItem>
                    <SystemPurposeCard
                      activationKey={activationKey}
                      actionHandler={handleEditActivationKeyModalToggle}
                    />
                  </GalleryItem>
                  <GalleryItem>
                    <WorkloadCard
                      activationKey={activationKey}
                      actionHandler={handleEditReleaseVersionModalToggle}
                    />
                  </GalleryItem>
                </Gallery>
              </GridItem>
              <GridItem span={12}>
                <AdditionalRepositoriesCard
                  activationKey={activationKey}
                  actionHandler={handleEditActivationKeyModalToggle}
                />
              </GridItem>
            </Grid>
          </Main>
        </React.Fragment>
      )}
      <DeleteActivationKeyConfirmationModal
        handleModalToggle={handleDeleteActivationKeyModalToggle}
        isOpen={isDeleteActivationKeyModalOpen}
        name={id}
      />
      <EditActivationKeyModal
        title="Edit system purpose"
        description={editModalDescription}
        isOpen={isEditActivationKeyModalOpen}
        handleModalToggle={handleEditActivationKeyModalToggle}
        activationKeyName={id}
        systemPurposeOnly={true}
        modalSize="small"
      />
      <EditReleaseVersionModal
        isOpen={isEditReleaseVersionModalOpen}
        onClose={handleEditReleaseVersionModalToggle}
        releaseVersions={releaseVersions}
        areReleaseVersionsLoading={areReleaseVersionsLoading}
        activationKey={activationKey}
      />
    </React.Fragment>
  );
};

export default ActivationKey;
