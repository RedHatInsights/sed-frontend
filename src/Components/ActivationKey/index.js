import ActivationKey from './ActivationKey';

export default ActivationKey;
