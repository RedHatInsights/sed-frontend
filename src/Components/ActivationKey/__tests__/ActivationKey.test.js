import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import ActivationKey from '../index';
import { Provider } from 'react-redux';
import Authentication from '../../../Components/Authentication';
import { BrowserRouter as Router } from 'react-router-dom';
import { init } from '../../../store';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import useUser from '../../../hooks/useUser';
import { get, def } from 'bdd-lazy-var/global';
import useActivationKey from '../../../hooks/useActivationKey';
import '@testing-library/jest-dom';
import useAvailableRepositories from '../../../hooks/useAvailableRepositories';

jest.mock('../../../hooks/useAvailableRepositories');
jest.mock('../../../hooks/useActivationKey');
jest.mock('../../../hooks/useUser');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '/',
  }),
}));

const queryClient = new QueryClient();

const PageContainer = () => (
  <QueryClientProvider client={queryClient}>
    <Authentication>
      <Provider store={init().getStore()}>
        <Router>
          <ActivationKey />
        </Router>
      </Provider>
    </Authentication>
  </QueryClientProvider>
);

const mockAuthenticateUser = (isLoading, isError, rbacPermissions) => {
  const user = {
    accountNumber: '123',
    orgId: '123',
    rbacPermissions: rbacPermissions,
  };
  useUser.mockReturnValue({
    isLoading: isLoading,
    isFetching: false,
    isSuccess: true,
    isError: isError,
    data: user,
  });
  if (isError === false) {
    queryClient.setQueryData(['user'], user);
  }
};

// eslint-disable-next-line react/display-name
jest.mock('../../../Components/AdditionalRepositoriesTable', () => () => (
  <div>Additional Repositories Table</div>
));
// eslint-disable-next-line react/display-name
jest.mock('../../../Components/shared/breadcrumbs', () => () => (
  <div>Breadcrumbs</div>
));

jest.mock(
  '@redhat-cloud-services/frontend-components/NotAuthorized',
  // eslint-disable-next-line react/display-name
  () => () => <div>Not Authorized</div>
);

jest.mock(
  '@redhat-cloud-services/frontend-components/Unavailable',
  // eslint-disable-next-line react/display-name
  () => () => <div>Unavailable</div>
);

describe('ActivationKey', () => {
  def('isLoading', () => false);
  def('isError', () => false);
  def('rbacPermissions', () => {
    return { canReadActivationKeys: true, canWriteActivationKeys: true };
  });
  def('keyData', () => {
    return {
      name: 'A',
      role: 'B',
      serviceLevel: 'C',
      usage: 'D',
      additionalRepositories: [],
      releaseVersion: 1,
    };
  });

  beforeEach(() => {
    jest.resetAllMocks();
    mockAuthenticateUser(
      get('isLoading'),
      get('isError'),
      get('rbacPermissions')
    );
    useActivationKey.mockReturnValue({
      isLoading: false,
      isFetching: false,
      isError: false,
      isSuccess: true,
      data: get('keyData'),
    });
  });

  it('renders correctly', async () => {
    useAvailableRepositories.mockReturnValue({
      isLoading: false,
      isFetching: false,
      isError: false,
      isSuccess: true,
      data: [], // Mock the data that the hook returns
    });

    render(
      <QueryClientProvider client={queryClient}>
        <PageContainer />
      </QueryClientProvider>
    );

    await waitFor(() =>
      expect(screen.getByText('System Purpose')).toBeInTheDocument()
    );
    expect(screen.getByText('Workload')).toBeInTheDocument();
    expect(screen.getByText('Additional repositories')).toBeInTheDocument();
  });

  describe('when the user call fails', () => {
    def('isError', () => true);

    it('renders an error message when user call fails', async () => {
      const { container } = render(<PageContainer />);
      await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
      expect(container).toMatchSnapshot();
    });
  });

  describe('when the user does not have proper permissions', () => {
    def('rbacPermissions', () => {
      return { canReadActivationKeys: false };
    });

    it('redirects to not authorized page', async () => {
      render(<PageContainer />);
      await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
      expect(screen.getByText('Not Authorized')).toBeInTheDocument();
    });
  });
});
