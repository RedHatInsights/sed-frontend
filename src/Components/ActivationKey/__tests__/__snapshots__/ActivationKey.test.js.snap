// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivationKey when the user call fails renders an error message when user call fails 1`] = `
<div>
  <div>
    Unavailable
  </div>
</div>
`;
